lambda-runtime/src/main/scala/feral/lambda/runtime/LambdaRuntime.scala

```diff
@@ -0,0 +1,94 @@
+    handler.use(handler => env.lambdaRuntimeApi.flatMap(api => processEvents(api, client, handler)))
+
+  private def processEvents[F[_]](runtimeApi: String, client: Client[F], handlerFun: (Json, Context[F]) => F[Json])(implicit F: Temporal[F], env: LambdaRuntimeEnv[F]): F[Unit] = {
+    implicit val jsonEncoder: EntityEncoder[F, Json] = jsonEncoderWithPrinter[F](Printer.noSpaces.copy(dropNullValues = true))
+    (for {
+      nextInvocationUrl <- getNextInvocationUrl(runtimeApi)
+      request <- client.get(nextInvocationUrl)(LambdaRequest.fromResponse[F])
+      invocationErrorUrl <- getInvocationErrorUrl(runtimeApi, request.id) // unsure if errors in these lines can be counted as initialization errors or not, runtime crashes for just now
+      _ <- (for {
+        context <- createContext(request)
+        handlerFiber <- handlerFun(request.body, context).start
+        result <- handlerFiber.join.flatMap {
+          case Succeeded(result: F[Json]) => result
+          case Errored(e: Throwable) => F.raiseError[Json](e)
+          case Canceled() => F.raiseError[Json](new CancellationException)
+        }
+        invocationResponseUrl <- getInvocationResponseUrl(runtimeApi, request.id)
+        _ <- client.expect[Unit](Request(POST, invocationResponseUrl).withEntity(result))
+      } yield ()).handleErrorWith(e => {
+          val error = LambdaErrorRequest(e.getMessage, "exception", List())
+          client.expect[Unit](Request(POST, invocationErrorUrl).withEntity(error.asJson))
+        })
+    } yield ()).foreverM
+  }
+  
+  private def createContext[F[_]](request: LambdaRequest)(implicit F: Temporal[F], env: LambdaRuntimeEnv[F]): F[Context[F]] = for {
+    functionName <- env.lambdaFunctionName
+    functionVersion <- env.lambdaFunctionVersion
+    functionMemorySize <- env.lambdaFunctionMemorySize
+    logGroupName <- env.lambdaLogGroupName
+    logStreamName <- env.lambdaLogStreamName
+  } yield {
+    new Context[F](
+      functionName,
+      functionVersion,
+      request.invokedFunctionArn,
+      functionMemorySize,
+      request.id,
+      logGroupName,
+      logStreamName,
+      request.identity,
+      request.clientContext,
+      F.realTime.map(curTime => request.deadlineTimeInMs - curTime)
+    )
+  }
+
+  private def getNextInvocationUrl[F[_]: MonadThrow](api: String): F[Uri] = Uri.fromString(s"/$api/$ApiVersion/runtime/invocation/next").liftTo[F]
+
+  private def getInvocationResponseUrl[F[_]: MonadThrow](api: String, id: String): F[Uri] = Uri.fromString(s"/$api/$ApiVersion/runtime/invocation/$id/response").liftTo[F]
+
+  private def getInitErrorUrl[F[_]: MonadThrow](api: String): F[Uri] = Uri.fromString(s"/$api/$ApiVersion/runtime/init/error").liftTo[F]
+
+  private def getInvocationErrorUrl[F[_]: MonadThrow](api: String, requestId: String): F[Uri] = Uri.fromString(s"/$api/$ApiVersion/runtime/invocation/$requestId/error").liftTo[F]
```

Comment:
I think we can do `Uri.fromString(s"/$api/$ApiVersion/runtime/")` once, and then safely all the other URIs using that.